In [705]:
# Supress Warnings

import warnings
warnings.filterwarnings('ignore')

In [706]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

## Reading and inspecting 'companies.txt'

In [707]:
# Importing 'companies.txt' into companies dataframe
companies = pd.read_csv('companies.txt', delimiter ='\t', encoding ='ISO-8859-1')

In [708]:
# Read companies dataframe
companies.head()

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [709]:
# Shape of companies dataframe
companies.shape

(66368, 10)

In [710]:
# Information about the dataframe
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
permalink        66368 non-null object
name             66367 non-null object
homepage_url     61310 non-null object
category_list    63220 non-null object
status           66368 non-null object
country_code     59410 non-null object
state_code       57821 non-null object
region           58338 non-null object
city             58340 non-null object
founded_at       51147 non-null object
dtypes: object(10)
memory usage: 5.1+ MB


In [711]:
# As mentioned in the problem statement encoding issues exist and are identified
companies.tail()

permalink  \
66363  /Organization/Zznode-Science-And-Technology-Co...   
66364                          /Organization/Zzzzapp-Com   
66365                               /Organization/ÁEron   
66366                             /Organization/ÔAsys-2   
66367  /Organization/Ä°Novatiff-Reklam-Ve-Tanä±Tä±M-H...   

                                                name            homepage_url  \
66363                  ZZNode Science and Technology   http://www.zznode.com   
66364                          Zzzzapp Wireless ltd.  http://www.zzzzapp.com   
66365                                         ÁERON    http://www.aeron.hu/   
66366                                         Ôasys    http://www.oasys.io/   
66367  İnovatiff Reklam ve Tanıtım Hizmetleri Tic    http://inovatiff.com   

                                           category_list     status  \
66363                                Enterprise Software  operating   
66364        Advertising|Mobile|Web Development|Wireless  operating   
66365                                                NaN  operating   
66366  Consumer Electronics|Internet of Things|Teleco...  operating   
66367                 Consumer Goods|E-Commerce|Internet  operating   

      country_code state_code       region           city  founded_at  
66363          CHN         22      Beijing        Beijing         NaN  
66364          HRV         15        Split          Split  13-05-2012  
66365          NaN        NaN          NaN            NaN  01-01-2011  
66366          USA         CA  SF Bay Area  San Francisco  01-01-2014  
66367          NaN        NaN          NaN            NaN         NaN

In [712]:
# Handling encoding issues in permalink and name column
companies['permalink'] = companies.permalink.str.encode('utf-8').str.decode('ascii', 'ignore')
companies.tail()

permalink  \
66363  /Organization/Zznode-Science-And-Technology-Co...   
66364                          /Organization/Zzzzapp-Com   
66365                                 /Organization/Eron   
66366                               /Organization/Asys-2   
66367  /Organization/Novatiff-Reklam-Ve-TanTM-Hizmetl...   

                                                name            homepage_url  \
66363                  ZZNode Science and Technology   http://www.zznode.com   
66364                          Zzzzapp Wireless ltd.  http://www.zzzzapp.com   
66365                                         ÁERON    http://www.aeron.hu/   
66366                                         Ôasys    http://www.oasys.io/   
66367  İnovatiff Reklam ve Tanıtım Hizmetleri Tic    http://inovatiff.com   

                                           category_list     status  \
66363                                Enterprise Software  operating   
66364        Advertising|Mobile|Web Development|Wireless  operating   
66365                                                NaN  operating   
66366  Consumer Electronics|Internet of Things|Teleco...  operating   
66367                 Consumer Goods|E-Commerce|Internet  operating   

      country_code state_code       region           city  founded_at  
66363          CHN         22      Beijing        Beijing         NaN  
66364          HRV         15        Split          Split  13-05-2012  
66365          NaN        NaN          NaN            NaN  01-01-2011  
66366          USA         CA  SF Bay Area  San Francisco  01-01-2014  
66367          NaN        NaN          NaN            NaN         NaN

In [713]:
# Converting permalink column to lower case
companies['permalink']=companies['permalink'].apply(lambda x: x.lower())
companies.permalink.head()

0                    /organization/-fame
1                 /organization/-qounter
2    /organization/-the-one-of-them-inc-
3                  /organization/0-6-com
4         /organization/004-technologies
Name: permalink, dtype: object

## Reading and inspecting 'rounds2.csv'

In [714]:
# Importing 'round2.csv' into rounds2 dataframe
rounds2 = pd.read_csv('rounds2.csv', encoding = 'ISO-8859-1')
rounds2.head()

company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [715]:
# Inspecting the shape of rounds2 dataframe
rounds2.shape

(114949, 6)

In [716]:
# Information about the rounds2 dataframe
rounds2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
company_permalink          114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB


In [717]:
# As mentioned in the problem statement encoding issues exist and are identified
rounds2.tail()

company_permalink  \
114944                          /organization/zzzzapp-com   
114945                          /ORGANIZATION/ZZZZAPP-COM   
114946                               /organization/ãeron   
114947                             /ORGANIZATION/ÔASYS-2   
114948  /organization/ä°novatiff-reklam-ve-tanä±tä±m-h...   

                                funding_round_permalink funding_round_type  \
114944  /funding-round/8f6d25b8ee4199e586484d817bceda05   convertible_note   
114945  /funding-round/ff1aa06ed5da186c84f101549035d4ae               seed   
114946  /funding-round/59f4dce44723b794f21ded3daed6e4fe            venture   
114947  /funding-round/35f09d0794651719b02bbfd859ba9ff5               seed   
114948  /funding-round/af942869878d2cd788ef5189b435ebc4              grant   

       funding_round_code   funded_at  raised_amount_usd  
114944                NaN  01-03-2014            41313.0  
114945                NaN  01-05-2013            32842.0  
114946                  A  01-08-2014                NaN  
114947                NaN  01-01-2015            18192.0  
114948                NaN  01-10-2013            14851.0

In [718]:
# Handling encoding issues in company_permalink column
rounds2['company_permalink'] = rounds2['company_permalink'].str.encode('utf-8').str.decode('ascii', 'ignore')
companies.tail()

permalink  \
66363  /organization/zznode-science-and-technology-co...   
66364                          /organization/zzzzapp-com   
66365                                 /organization/eron   
66366                               /organization/asys-2   
66367  /organization/novatiff-reklam-ve-tantm-hizmetl...   

                                                name            homepage_url  \
66363                  ZZNode Science and Technology   http://www.zznode.com   
66364                          Zzzzapp Wireless ltd.  http://www.zzzzapp.com   
66365                                         ÁERON    http://www.aeron.hu/   
66366                                         Ôasys    http://www.oasys.io/   
66367  İnovatiff Reklam ve Tanıtım Hizmetleri Tic    http://inovatiff.com   

                                           category_list     status  \
66363                                Enterprise Software  operating   
66364        Advertising|Mobile|Web Development|Wireless  operating   
66365                                                NaN  operating   
66366  Consumer Electronics|Internet of Things|Teleco...  operating   
66367                 Consumer Goods|E-Commerce|Internet  operating   

      country_code state_code       region           city  founded_at  
66363          CHN         22      Beijing        Beijing         NaN  
66364          HRV         15        Split          Split  13-05-2012  
66365          NaN        NaN          NaN            NaN  01-01-2011  
66366          USA         CA  SF Bay Area  San Francisco  01-01-2014  
66367          NaN        NaN          NaN            NaN         NaN

In [719]:
# Converting company_permalink column to lower case
rounds2['company_permalink']=rounds2['company_permalink'].apply(lambda x: x.lower())
rounds2.company_permalink.head()

0                    /organization/-fame
1                 /organization/-qounter
2                 /organization/-qounter
3    /organization/-the-one-of-them-inc-
4                  /organization/0-6-com
Name: company_permalink, dtype: object

## Tabe 1.1: Understanding the Data Set

### Task 1: How many unique companies are present in rounds2?

In [720]:
# Finding unique companies in rounds2 dataframe
len(rounds2.company_permalink.unique().tolist())

66368

We have 66368 unique companies in rounds2 dataframe.

### Task2: How many unique companies are present in the companies file?

In [721]:
# Finding unique companies in companies dataframe
len(companies.permalink.unique().tolist())

66368

We have also got 66368 unique companies in the companies dataframe.

### Task3: In the companies data frame, which column can be used as the  unique key for each company?

Permalink is the unique ID of a company. So, the permalink column can be used to identify each company.

### Task4: Are there any companies in the rounds2 file which are not  present in companies ?

In [722]:
# Verifying if all the permalinks in rounds2 dataframe are a subset of the permalinks in comapanies dataframe
rounds2['company_permalink'].isin(companies['permalink']).value_counts()

True    114949
Name: company_permalink, dtype: int64

The above result goes to say that all the permalinks in rounds2 dataframe exist in the companies dataframe. So, there are **NO** companies in rounds2 file which are not a part of companies file.

In [723]:
# Rename company_permalink of rounds2 dataframe to permalink for the purpose of merging
rounds2 = rounds2.rename(columns={'company_permalink': 'permalink'})

In [724]:
# Creating a master_frame by merging companies dataframe with rounds2 dataframe
master_frame = pd.merge(rounds2, companies, on = 'permalink', how = 'left')
master_frame.head()

permalink  \
0                  /organization/-fame   
1               /organization/-qounter   
2               /organization/-qounter   
3  /organization/-the-one-of-them-inc-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd                    name  \
0                  B  05-01-2015         10000000.0                   #fame   
1                  A  14-10-2014                NaN                :Qounter   
2                NaN  01-03-2014           700000.0                :Qounter   
3                  B  30-01-2014          3406878.0  (THE) ONE of THEM,Inc.   
4                  A  19-03-2008          2000000.0                 0-6.com   

             homepage_url                                      category_list  \
0      http://livfame.com                                              Media   
1  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
2  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
3     http://oneofthem.jp                                  Apps|Games|Mobile   
4      http://www.0-6.com                                        Curated Web   

      status country_code state_code      region           city  founded_at  
0  operating          IND         16      Mumbai         Mumbai         NaN  
1  operating          USA         DE  DE - Other  Delaware City  04-09-2014  
2  operating          USA         DE  DE - Other  Delaware City  04-09-2014  
3  operating          NaN        NaN         NaN            NaN         NaN  
4  operating          CHN         22     Beijing        Beijing  01-01-2007

In [725]:
#Understanding the master_frame dataframe
master_frame.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 114949 entries, 0 to 114948
Data columns (total 15 columns):
permalink                  114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
name                       114948 non-null object
homepage_url               108815 non-null object
category_list              111539 non-null object
status                     114949 non-null object
country_code               106271 non-null object
state_code                 104003 non-null object
region                     104782 non-null object
city                       104785 non-null object
founded_at                 94428 non-null object
dtypes: float64(1), object(14)
memory usage: 14.0+ MB


### Task5: How many observations are present in master_frame ?

In [726]:
# Finding the total number of observations in master_frame
len(master_frame.index)

114949

We have 114949 observations which is same as the rounds2 data frame. It suggests that merge has taken place successfully.

## Cleaning master_frame dataframe

In [727]:
# Column wise null count
master_frame.isnull().sum()

permalink                      0
funding_round_permalink        0
funding_round_type             0
funding_round_code         83809
funded_at                      0
raised_amount_usd          19990
name                           1
homepage_url                6134
category_list               3410
status                         0
country_code                8678
state_code                 10946
region                     10167
city                       10164
founded_at                 20521
dtype: int64

In [728]:
# Dropping columns irrelavant to the analysis
master_frame.drop(['funding_round_permalink', 'funding_round_code', 'funded_at', 'homepage_url', 'state_code', 'region','city',
                  'founded_at'], axis=1, inplace = True)
master_frame.head()

permalink funding_round_type  raised_amount_usd  \
0                  /organization/-fame            venture         10000000.0   
1               /organization/-qounter            venture                NaN   
2               /organization/-qounter               seed           700000.0   
3  /organization/-the-one-of-them-inc-            venture          3406878.0   
4                /organization/0-6-com            venture          2000000.0   

                     name                                      category_list  \
0                   #fame                                              Media   
1                :Qounter  Application Platforms|Real Time|Social Network...   
2                :Qounter  Application Platforms|Real Time|Social Network...   
3  (THE) ONE of THEM,Inc.                                  Apps|Games|Mobile   
4                 0-6.com                                        Curated Web   

      status country_code  
0  operating          IND  
1  operating          USA  
2  operating          USA  
3  operating          NaN  
4  operating          CHN

In [729]:
# Finding column-wise null percentages
round(100*(master_frame.isnull().sum()/len(master_frame.index)),4)

permalink              0.0000
funding_round_type     0.0000
raised_amount_usd     17.3903
name                   0.0009
category_list          2.9665
status                 0.0000
country_code           7.5494
dtype: float64

In [730]:
# Null percentage of raised_amount_usd is too high to impute values. So, we will remove the rows which have null values in raised_amount_usd
master_frame = master_frame[~np.isnan(master_frame['raised_amount_usd'])]

In [731]:
# Finding column-wise null percentages
round(100*(master_frame.isnull().sum()/len(master_frame.index)),4)

permalink             0.0000
funding_round_type    0.0000
raised_amount_usd     0.0000
name                  0.0011
category_list         1.0994
status                0.0000
country_code          6.1616
dtype: float64

In [732]:
# Removing rows of the country code column which are null
master_frame = master_frame[~pd.isnull(master_frame['country_code'])]

In [733]:
# Finding column-wise null percentages
round(100*(master_frame.isnull().sum()/len(master_frame.index)),4)

permalink             0.0000
funding_round_type    0.0000
raised_amount_usd     0.0000
name                  0.0011
category_list         0.6498
status                0.0000
country_code          0.0000
dtype: float64

In [734]:
# Removing null values from category_list as it is not practical to impute
master_frame = master_frame[~pd.isnull(master_frame['category_list'])]

In [735]:
# Finding column-wise null percentages
round(100*(master_frame.isnull().sum()/len(master_frame.index)),4)

permalink             0.0000
funding_round_type    0.0000
raised_amount_usd     0.0000
name                  0.0011
category_list         0.0000
status                0.0000
country_code          0.0000
dtype: float64

In [736]:
# Removing the null values in the name column
master_frame = master_frame[~pd.isnull(master_frame['name'])]

In [737]:
# Number of firms with zero funding
master_frame[(master_frame['raised_amount_usd']==0)].shape

(363, 7)

In [738]:
# Removing companies with zero funding. It wouldn't affect the analysis as they are not present in big numbers(only 363)
master_frame = master_frame[(master_frame['raised_amount_usd']!=0)]

In [739]:
# Making sure we have handled all the null values
round(100*(master_frame.isnull().sum()/len(master_frame.index)),2)

permalink             0.0
funding_round_type    0.0
raised_amount_usd     0.0
name                  0.0
category_list         0.0
status                0.0
country_code          0.0
dtype: float64

In [740]:
# Shape of master_frame
master_frame.shape

(88165, 7)

We have successfully handled null values in the master_frame dataframe and ended up with 88166 data points to work with.

In [741]:
#Pecentage of rows retained
round((len(master_frame.index)/114949)*100,2)

76.7

In [742]:
master_frame.head()

permalink funding_round_type  raised_amount_usd  \
0                  /organization/-fame            venture         10000000.0   
2               /organization/-qounter               seed           700000.0   
4                /organization/0-6-com            venture          2000000.0   
6     /organization/01games-technology        undisclosed            41250.0   
7  /organization/0ndine-biomedical-inc               seed            43360.0   

                     name                                      category_list  \
0                   #fame                                              Media   
2                :Qounter  Application Platforms|Real Time|Social Network...   
4                 0-6.com                                        Curated Web   
6      01Games Technology                                              Games   
7  Ondine Biomedical Inc.                                      Biotechnology   

      status country_code  
0  operating          IND  
2  operating          USA  
4  operating          CHN  
6  operating          HKG  
7  operating          CAN

## Table 2.1:  Average Values of Investments for Each of the Funding Types

In [743]:
# Finding the average value of investments for each of the funding type
group_fund = master_frame.groupby('funding_round_type')
group_fund.raised_amount_usd.mean()

funding_round_type
angel                    9.775726e+05
convertible_note         1.360899e+06
debt_financing           1.722590e+07
equity_crowdfunding      6.028982e+05
grant                    4.517358e+06
non_equity_assistance    5.060562e+05
post_ipo_debt            1.694518e+08
post_ipo_equity          6.607706e+07
private_equity           7.397913e+07
product_crowdfunding     1.365642e+06
secondary_market         8.443853e+07
seed                     7.503536e+05
undisclosed              1.605882e+07
venture                  1.172987e+07
Name: raised_amount_usd, dtype: float64

### Task 1: Average funding amount for venture type

\\$11.72987e+06 or \\$11.72 M

### Task 2: Average funding amount of angel type

\\$9.775726e+05 or \\$0.98 M

### Task 3: Average funding amount of seed type

\\$7.503191e+05 or \\$0.75 M

### Task 4: Average funding amount of private equity type

\\$73.97913e+06 or \\$73.98M

### Task 5: Considering that Spark Funds wants to invest between 5 to 15 million USD per  investment round, which investment type is the most suitable for them?

Venture type investment is the most sutable investment type

In [744]:
# Let's narrow down our master_frame to concentrate on investments of venture type
master_frame = master_frame.loc[master_frame['funding_round_type']=='venture']
master_frame.head()

permalink funding_round_type  raised_amount_usd  \
0                   /organization/-fame            venture         10000000.0   
4                 /organization/0-6-com            venture          2000000.0   
8   /organization/0ndine-biomedical-inc            venture           719491.0   
10                 /organization/0xdata            venture         20000000.0   
11                 /organization/0xdata            venture          1700000.0   

                      name  category_list     status country_code  
0                    #fame          Media  operating          IND  
4                  0-6.com    Curated Web  operating          CHN  
8   Ondine Biomedical Inc.  Biotechnology  operating          CAN  
10                  H2O.ai      Analytics  operating          USA  
11                  H2O.ai      Analytics  operating          USA

In [745]:
# Country-wise analysis of venture investment
group_country = master_frame.groupby('country_code')

In [746]:
# Creating a dataframe top9 which contains the top 9 countries which are best for venture investment
top9 = pd.DataFrame(group_country.raised_amount_usd.sum()).sort_values('raised_amount_usd',ascending = False).head(9)
top9

raised_amount_usd
country_code                   
USA                4.200680e+11
CHN                3.933892e+10
GBR                2.007281e+10
IND                1.426151e+10
CAN                9.482218e+09
FRA                7.226851e+09
ISR                6.854350e+09
DEU                6.306922e+09
JPN                3.167647e+09

## Table 3.1: Analysing the Top 3 English-Speaking Countries in top9 dataframe

1. USA
2. GBR
3. IND

## Preparing master_frame for sector analysis

In [747]:
# Creating a new column primary sector and extract from category_list column
split_category = master_frame['category_list'].str.split('|',-1)
master_frame['primary_sector']=split_category.str.get(0)
master_frame.head()

permalink funding_round_type  raised_amount_usd  \
0                   /organization/-fame            venture         10000000.0   
4                 /organization/0-6-com            venture          2000000.0   
8   /organization/0ndine-biomedical-inc            venture           719491.0   
10                 /organization/0xdata            venture         20000000.0   
11                 /organization/0xdata            venture          1700000.0   

                      name  category_list     status country_code  \
0                    #fame          Media  operating          IND   
4                  0-6.com    Curated Web  operating          CHN   
8   Ondine Biomedical Inc.  Biotechnology  operating          CAN   
10                  H2O.ai      Analytics  operating          USA   
11                  H2O.ai      Analytics  operating          USA   

   primary_sector  
0           Media  
4     Curated Web  
8   Biotechnology  
10      Analytics  
11      Analytics

## Sector Analysis 1: Determine main sector for each of the primary sector in master_frame

In [748]:
# Import mapping.csv into mapping dataframe
mapping=pd.read_csv('mapping.csv', encoding='latin1')
mapping.head()

category_list  Automotive & Sports  Blanks  Cleantech / Semiconductors  \
0            NaN                    0       1                           0   
1             3D                    0       0                           0   
2    3D Printing                    0       0                           0   
3  3D Technology                    0       0                           0   
4     Accounting                    0       0                           0   

   Entertainment  Health  Manufacturing  News, Search and Messaging  Others  \
0              0       0              0                           0       0   
1              0       0              1                           0       0   
2              0       0              1                           0       0   
3              0       0              1                           0       0   
4              0       0              0                           0       0   

   Social, Finance, Analytics, Advertising  
0                                        0  
1                                        0  
2                                        0  
3                                        0  
4                                        1

In [749]:
# Identified an error in category_list column
mapping[(mapping['category_list']!='nan') & (mapping['category_list'].str.contains('0', regex = False))].head()

category_list  Automotive & Sports  Blanks  \
20  Alter0tive Medicine                    0       0   
22             A0lytics                    0       0   
59    Big Data A0lytics                    0       0   
79    Business A0lytics                    0       0   
89              Can0bis                    0       0   

    Cleantech / Semiconductors  Entertainment  Health  Manufacturing  \
20                           0              0       1              0   
22                           0              0       0              0   
59                           0              0       0              0   
79                           0              0       0              0   
89                           0              0       1              0   

    News, Search and Messaging  Others  \
20                           0       0   
22                           0       0   
59                           0       0   
79                           0       0   
89                           0       0   

    Social, Finance, Analytics, Advertising  
20                                        0  
22                                        1  
59                                        1  
79                                        1  
89                                        0

In [750]:
# Handling error in  category_list column. '0' to be replaced with 'na'.
mapping['category_list']=mapping['category_list'].str.replace('0','na')
mapping[(mapping['category_list']!='nan') & (mapping['category_list'].str.contains('na', regex = False))].head()

category_list  Automotive & Sports  Blanks  \
20  Alternative Medicine                    0       0   
22             Analytics                    0       0   
59    Big Data Analytics                    0       0   
79    Business Analytics                    0       0   
89              Cannabis                    0       0   

    Cleantech / Semiconductors  Entertainment  Health  Manufacturing  \
20                           0              0       1              0   
22                           0              0       0              0   
59                           0              0       0              0   
79                           0              0       0              0   
89                           0              0       1              0   

    News, Search and Messaging  Others  \
20                           0       0   
22                           0       0   
59                           0       0   
79                           0       0   
89                           0       0   

    Social, Finance, Analytics, Advertising  
20                                        0  
22                                        1  
59                                        1  
79                                        1  
89                                        0

In [751]:
# We have a special case in category_list where one of the primary sector is enterptise 2.0. Need to handle this exception.
mapping['category_list']=mapping['category_list'].str.replace('Enterprise 2.na','Enterprise 2.0')

In [752]:
# Unpivoting dataframe from short format to long format
mapping = (pd.melt(mapping, id_vars='category_list', var_name='map_sector'))
mapping.head(10)

category_list           map_sector  value
0                 NaN  Automotive & Sports      0
1                  3D  Automotive & Sports      0
2         3D Printing  Automotive & Sports      0
3       3D Technology  Automotive & Sports      0
4          Accounting  Automotive & Sports      0
5    Active Lifestyle  Automotive & Sports      0
6        Ad Targeting  Automotive & Sports      0
7  Advanced Materials  Automotive & Sports      0
8    Adventure Travel  Automotive & Sports      1
9         Advertising  Automotive & Sports      0

In [753]:
mapping['category_list']=mapping['category_list'].apply(lambda x: str(x).lower())
mapping.head()

category_list           map_sector  value
0            nan  Automotive & Sports      0
1             3d  Automotive & Sports      0
2    3d printing  Automotive & Sports      0
3  3d technology  Automotive & Sports      0
4     accounting  Automotive & Sports      0

In [754]:
# Filter value column for 1's to determine main sector corresponding to each primary sector in categpory_list
mapping = mapping[mapping['value']==1]
mapping.head()

category_list           map_sector  value
8   adventure travel  Automotive & Sports      1
14         aerospace  Automotive & Sports      1
45              auto  Automotive & Sports      1
46   automated kiosk  Automotive & Sports      1
47        automotive  Automotive & Sports      1

In [755]:
# Creating a dictionary which handles key, value pairs of category_list & map_sector respectively
dict1 = mapping.set_index('category_list')['map_sector'].to_dict()
dict1

{'adventure travel': 'Automotive & Sports',
 'aerospace': 'Automotive & Sports',
 'auto': 'Automotive & Sports',
 'automated kiosk': 'Automotive & Sports',
 'automotive': 'Automotive & Sports',
 'bicycles': 'Automotive & Sports',
 'boating industry': 'Automotive & Sports',
 'cad': 'Automotive & Sports',
 'cars': 'Automotive & Sports',
 'design': 'Automotive & Sports',
 'designers': 'Automotive & Sports',
 'drones': 'Automotive & Sports',
 'electric vehicles': 'Automotive & Sports',
 'fuels': 'Automotive & Sports',
 'heavy industry': 'Automotive & Sports',
 'industrial automation': 'Automotive & Sports',
 'limousines': 'Automotive & Sports',
 'soccer': 'Automotive & Sports',
 'sporting goods': 'Automotive & Sports',
 'sports': 'Automotive & Sports',
 'transportation': 'Automotive & Sports',
 'unmanned air systems': 'Automotive & Sports',
 'nan': 'Blanks',
 'air pollution control': 'Cleantech / Semiconductors',
 'aquaculture': 'Cleantech / Semiconductors',
 'bio-pharm': 'Cleantech / Semi

In [756]:
# Converting main_sector column of master_frame to lower case
master_frame['primary_sector']=master_frame['primary_sector'].apply(lambda x: str(x).lower())
master_frame.head()

permalink funding_round_type  raised_amount_usd  \
0                   /organization/-fame            venture         10000000.0   
4                 /organization/0-6-com            venture          2000000.0   
8   /organization/0ndine-biomedical-inc            venture           719491.0   
10                 /organization/0xdata            venture         20000000.0   
11                 /organization/0xdata            venture          1700000.0   

                      name  category_list     status country_code  \
0                    #fame          Media  operating          IND   
4                  0-6.com    Curated Web  operating          CHN   
8   Ondine Biomedical Inc.  Biotechnology  operating          CAN   
10                  H2O.ai      Analytics  operating          USA   
11                  H2O.ai      Analytics  operating          USA   

   primary_sector  
0           media  
4     curated web  
8   biotechnology  
10      analytics  
11      analytics

In [757]:
# Creating column main_sector in master_frame dataframe and use the values in dict1 to fill the column using map function
master_frame['main_sector'] = master_frame['primary_sector'].map(dict1)
master_frame.head()

permalink funding_round_type  raised_amount_usd  \
0                   /organization/-fame            venture         10000000.0   
4                 /organization/0-6-com            venture          2000000.0   
8   /organization/0ndine-biomedical-inc            venture           719491.0   
10                 /organization/0xdata            venture         20000000.0   
11                 /organization/0xdata            venture          1700000.0   

                      name  category_list     status country_code  \
0                    #fame          Media  operating          IND   
4                  0-6.com    Curated Web  operating          CHN   
8   Ondine Biomedical Inc.  Biotechnology  operating          CAN   
10                  H2O.ai      Analytics  operating          USA   
11                  H2O.ai      Analytics  operating          USA   

   primary_sector                              main_sector  
0           media                            Entertainment  
4     curated web               News, Search and Messaging  
8   biotechnology               Cleantech / Semiconductors  
10      analytics  Social, Finance, Analytics, Advertising  
11      analytics  Social, Finance, Analytics, Advertising

In [758]:
# Identifying null values in the master_frame dataframe
master_frame.isnull().sum()

permalink              0
funding_round_type     0
raised_amount_usd      0
name                   0
category_list          0
status                 0
country_code           0
primary_sector         0
main_sector           13
dtype: int64

In [759]:
# Removing rows having null values
master_frame = master_frame[~pd.isnull(master_frame.main_sector)]
master_frame.head()

permalink funding_round_type  raised_amount_usd  \
0                   /organization/-fame            venture         10000000.0   
4                 /organization/0-6-com            venture          2000000.0   
8   /organization/0ndine-biomedical-inc            venture           719491.0   
10                 /organization/0xdata            venture         20000000.0   
11                 /organization/0xdata            venture          1700000.0   

                      name  category_list     status country_code  \
0                    #fame          Media  operating          IND   
4                  0-6.com    Curated Web  operating          CHN   
8   Ondine Biomedical Inc.  Biotechnology  operating          CAN   
10                  H2O.ai      Analytics  operating          USA   
11                  H2O.ai      Analytics  operating          USA   

   primary_sector                              main_sector  
0           media                            Entertainment  
4     curated web               News, Search and Messaging  
8   biotechnology               Cleantech / Semiconductors  
10      analytics  Social, Finance, Analytics, Advertising  
11      analytics  Social, Finance, Analytics, Advertising

## Sector Analysis 2: Creating data frames D1, D2 & D3.

### Creating D1

In [760]:
# Creating data frame D1 where country = USA & $5 M<=funding<=& 15 M
D1 = master_frame[(master_frame.country_code == 'USA') & (master_frame.raised_amount_usd>=5000000) & 
                  (master_frame.raised_amount_usd<=15000000)]
D1.head()

permalink funding_round_type  raised_amount_usd  \
12              /organization/0xdata            venture          8900000.0   
22        /organization/1-mainstream            venture          5000000.0   
103     /organization/128-technology            venture         11999347.0   
110  /organization/1366-technologies            venture         15000000.0   
111  /organization/1366-technologies            venture          5000000.0   

                  name                     category_list     status  \
12              H2O.ai                         Analytics  operating   
22        1 Mainstream  Apps|Cable|Distribution|Software   acquired   
103     128 Technology      Service Providers|Technology  operating   
110  1366 Technologies                     Manufacturing  operating   
111  1366 Technologies                     Manufacturing  operating   

    country_code     primary_sector                              main_sector  
12           USA          analytics  Social, Finance, Analytics, Advertising  
22           USA               apps               News, Search and Messaging  
103          USA  service providers                                   Others  
110          USA      manufacturing                            Manufacturing  
111          USA      manufacturing                            Manufacturing

In [761]:
# Adding new column 'num_invest_sector' & sum_invest_sector
D1['num_invest_sector'] = D1.groupby('main_sector')['raised_amount_usd'].transform('count')
D1['sum_invest_sector'] = D1.groupby('main_sector')['raised_amount_usd'].transform('sum')
D1.head()

permalink funding_round_type  raised_amount_usd  \
12              /organization/0xdata            venture          8900000.0   
22        /organization/1-mainstream            venture          5000000.0   
103     /organization/128-technology            venture         11999347.0   
110  /organization/1366-technologies            venture         15000000.0   
111  /organization/1366-technologies            venture          5000000.0   

                  name                     category_list     status  \
12              H2O.ai                         Analytics  operating   
22        1 Mainstream  Apps|Cable|Distribution|Software   acquired   
103     128 Technology      Service Providers|Technology  operating   
110  1366 Technologies                     Manufacturing  operating   
111  1366 Technologies                     Manufacturing  operating   

    country_code     primary_sector                              main_sector  \
12           USA          analytics  Social, Finance, Analytics, Advertising   
22           USA               apps               News, Search and Messaging   
103          USA  service providers                                   Others   
110          USA      manufacturing                            Manufacturing   
111          USA      manufacturing                            Manufacturing   

     num_invest_sector  sum_invest_sector  
12                2714       2.380738e+10  
22                1583       1.397157e+10  
103               2950       2.632101e+10  
110                799       7.258553e+09  
111                799       7.258553e+09

### Creating D2

In [762]:
# Creating data frame D2 where country = GBR & $5 M<=funding<=& 15 M
D2 = master_frame[(master_frame.country_code == 'GBR') & (master_frame.raised_amount_usd>=5000000) & 
                  (master_frame.raised_amount_usd<=15000000)]
D2.head()

permalink funding_round_type  raised_amount_usd       name  \
401  /organization/365scores            venture          5500000.0  365Scores   
864   /organization/7digital            venture          8468328.0   7digital   
865   /organization/7digital            venture         10000000.0   7digital   
942      /organization/90min            venture         15000000.0      90min   
943      /organization/90min            venture          5800000.0      90min   

                                         category_list     status  \
401                  Android|Apps|iPhone|Mobile|Sports  operating   
864  Content Creators|Content Delivery|Licensing|Mu...   acquired   
865  Content Creators|Content Delivery|Licensing|Mu...   acquired   
942                Media|News|Publishing|Soccer|Sports  operating   
943                Media|News|Publishing|Soccer|Sports  operating   

    country_code    primary_sector                              main_sector  
401          GBR           android  Social, Finance, Analytics, Advertising  
864          GBR  content creators                            Entertainment  
865          GBR  content creators                            Entertainment  
942          GBR             media                            Entertainment  
943          GBR             media                            Entertainment

In [763]:
# Adding new column 'num_invest_sector' & sum_invest_sector
D2['num_invest_sector'] = D2.groupby('main_sector')['raised_amount_usd'].transform('count')
D2['sum_invest_sector'] = D2.groupby('main_sector')['raised_amount_usd'].transform('sum')
D2.head()

permalink funding_round_type  raised_amount_usd       name  \
401  /organization/365scores            venture          5500000.0  365Scores   
864   /organization/7digital            venture          8468328.0   7digital   
865   /organization/7digital            venture         10000000.0   7digital   
942      /organization/90min            venture         15000000.0      90min   
943      /organization/90min            venture          5800000.0      90min   

                                         category_list     status  \
401                  Android|Apps|iPhone|Mobile|Sports  operating   
864  Content Creators|Content Delivery|Licensing|Mu...   acquired   
865  Content Creators|Content Delivery|Licensing|Mu...   acquired   
942                Media|News|Publishing|Soccer|Sports  operating   
943                Media|News|Publishing|Soccer|Sports  operating   

    country_code    primary_sector                              main_sector  \
401          GBR           android  Social, Finance, Analytics, Advertising   
864          GBR  content creators                            Entertainment   
865          GBR  content creators                            Entertainment   
942          GBR             media                            Entertainment   
943          GBR             media                            Entertainment   

     num_invest_sector  sum_invest_sector  
401                133       1.089404e+09  
864                 56       4.827847e+08  
865                 56       4.827847e+08  
942                 56       4.827847e+08  
943                 56       4.827847e+08

### Creating D3

In [764]:
# Creating data frame D3 where country = IND & $5 M<=funding<=& 15 M
D3 = master_frame[(master_frame.country_code == 'IND') & (master_frame.raised_amount_usd>=5000000) & 
                  (master_frame.raised_amount_usd<=15000000)]
D3.head()

permalink funding_round_type  raised_amount_usd  \
0                /organization/-fame            venture         10000000.0   
227   /organization/21diamonds-india            venture          6369507.0   
1049    /organization/a-little-world            venture          6410000.0   
2507    /organization/adlabs-imagica            venture          8180000.0   
3457             /organization/agile            venture          5740000.0   

                name                                 category_list     status  \
0              #fame                                         Media  operating   
227       21Diamonds                                    E-Commerce  operating   
1049  A LITTLE WORLD                                       Finance  operating   
2507  Adlabs Imagica                         Entertainment|Tourism  operating   
3457           Agile  Finance|Finance Technology|FinTech|Insurance  operating   

     country_code primary_sector                              main_sector  
0             IND          media                            Entertainment  
227           IND     e-commerce                                   Others  
1049          IND        finance  Social, Finance, Analytics, Advertising  
2507          IND  entertainment                            Entertainment  
3457          IND        finance  Social, Finance, Analytics, Advertising

In [765]:
# Adding new column 'num_invest_sector' & sum_invest_sector
D3['num_invest_sector'] = D3.groupby('main_sector')['raised_amount_usd'].transform('count')
D3['sum_invest_sector'] = D3.groupby('main_sector')['raised_amount_usd'].transform('sum')
D3.head()

permalink funding_round_type  raised_amount_usd  \
0                /organization/-fame            venture         10000000.0   
227   /organization/21diamonds-india            venture          6369507.0   
1049    /organization/a-little-world            venture          6410000.0   
2507    /organization/adlabs-imagica            venture          8180000.0   
3457             /organization/agile            venture          5740000.0   

                name                                 category_list     status  \
0              #fame                                         Media  operating   
227       21Diamonds                                    E-Commerce  operating   
1049  A LITTLE WORLD                                       Finance  operating   
2507  Adlabs Imagica                         Entertainment|Tourism  operating   
3457           Agile  Finance|Finance Technology|FinTech|Insurance  operating   

     country_code primary_sector                              main_sector  \
0             IND          media                            Entertainment   
227           IND     e-commerce                                   Others   
1049          IND        finance  Social, Finance, Analytics, Advertising   
2507          IND  entertainment                            Entertainment   
3457          IND        finance  Social, Finance, Analytics, Advertising   

      num_invest_sector  sum_invest_sector  
0                    33       2.808300e+08  
227                 110       1.013410e+09  
1049                 60       5.505496e+08  
2507                 33       2.808300e+08  
3457                 60       5.505496e+08

## Table 5.1: Sector-wise investment analysis

### Task 1: Total number of investments

In [766]:
# For the given constraints, total number of investments made in venture type in USA
D1.raised_amount_usd.count()

12063

In [767]:
# For the given constraints, total number of investments made in venture type in GBR
D2.raised_amount_usd.count()

621

In [768]:
# For the given constraints, total number of investments made in venture type in IND
D3.raised_amount_usd.count()

328

### Task 2: Total amount of investment 

In [769]:
# For the given constraints, total amount invested in venture type in USA
D1.raised_amount_usd.sum()

107757097294.0

In [770]:
# For the given constraints, total amount invested in venture type in GBR
D2.raised_amount_usd.sum()

5379078691.0

In [771]:
# For the given constraints, total amount invested in venture type in IND
D3.raised_amount_usd.sum()

2949543602.0

### Task 3,4,5,6,7,8: Total number of Investments (count)

In [772]:
# Top 3 sectors in D1 in terms of number of investments made
D1.groupby('main_sector')['num_invest_sector'].count().sort_values(ascending=False)

main_sector
Others                                     2950
Social, Finance, Analytics, Advertising    2714
Cleantech / Semiconductors                 2350
News, Search and Messaging                 1583
Health                                      909
Manufacturing                               799
Entertainment                               591
Automotive & Sports                         167
Name: num_invest_sector, dtype: int64

In [773]:
# Top 3 sectors in D2 in terms of number of investments made
D2.groupby('main_sector')['num_invest_sector'].count().sort_values(ascending=False)

main_sector
Others                                     147
Social, Finance, Analytics, Advertising    133
Cleantech / Semiconductors                 130
News, Search and Messaging                  73
Entertainment                               56
Manufacturing                               42
Health                                      24
Automotive & Sports                         16
Name: num_invest_sector, dtype: int64

In [774]:
# Top 3 sectors in D3 in terms of number of investments made
D3.groupby('main_sector')['num_invest_sector'].count().sort_values(ascending=False)

main_sector
Others                                     110
Social, Finance, Analytics, Advertising     60
News, Search and Messaging                  52
Entertainment                               33
Manufacturing                               21
Cleantech / Semiconductors                  20
Health                                      19
Automotive & Sports                         13
Name: num_invest_sector, dtype: int64

### Task 9: For top sector count-wise, which company received the highest investment ?

In [775]:
# For data frame D1
D1[D1.main_sector=='Others'].pivot_table(values='raised_amount_usd', 
                                         index = 'name', 
                                         aggfunc = 'sum' ).sort_values('raised_amount_usd',ascending=False).head()

raised_amount_usd
name                                  
Virtustream                 64300000.0
Capella Photonics           54968051.0
AirTight Networks           54201907.0
deCarta                     52100000.0
Black Duck Software         51000000.0

In [776]:
# For data frame D2
D2[D2.main_sector=='Others'].pivot_table(values='raised_amount_usd', 
                                         index = 'name', 
                                         aggfunc = 'sum' ).sort_values('raised_amount_usd',ascending=False).head()

raised_amount_usd
name                                      
Electric Cloud                  37000000.0
SenSage                         36250000.0
Enigmatec                       32500000.0
SilverRail Technologies         29000000.0
OpenCloud                       27972766.0

In [777]:
# For data frame D3
D3[D3.main_sector=='Others'].pivot_table(values='raised_amount_usd', 
                                         index = 'name', 
                                         aggfunc = 'sum' ).sort_values('raised_amount_usd',ascending=False).head()

raised_amount_usd
name                                
FirstCry.com              39000000.0
Myntra                    38000000.0
CommonFloor               32900000.0
Pepperfry.com             28000000.0
ItzCash Card Ltd.         25000000.0

### Task 10: For second best sector count-wise, which company received the highest investment ?

In [778]:
# For data frame D1
D1[D1.main_sector=='Social, Finance, Analytics, Advertising'].pivot_table(values='raised_amount_usd', index = 'name', 
                                                                         aggfunc = 'sum' ).sort_values('raised_amount_usd',
                                                                                                       ascending=False).head()

raised_amount_usd
name                                              
SST Inc. (Formerly ShotSpotter)         67933006.0
Demandbase                              63000000.0
Intacct                                 61800000.0
NetBase Solutions                       60600000.0
Lotame                                  59700000.0

In [779]:
# For data frame D2
D2[D2.main_sector=='Social, Finance, Analytics, Advertising'].pivot_table(values='raised_amount_usd', index = 'name', 
                                                                         aggfunc = 'sum' ).sort_values('raised_amount_usd',
                                                                                                       ascending=False).head()

raised_amount_usd
name                                    
Celltick Technologies         37500000.0
myThings                      34000000.0
Zopa                          32900000.0
VisualDNA                     28550000.0
MarketInvoice                 25553007.0

In [780]:
# For data frame D3
D3[D3.main_sector=='Social, Finance, Analytics, Advertising'].pivot_table(values='raised_amount_usd', index = 'name', 
                                                                         aggfunc = 'sum' ).sort_values('raised_amount_usd',
                                                                                                       ascending=False).head()

raised_amount_usd
name                                         
Manthan Systems                    50700000.0
Komli Media                        28000000.0
ShopClues.com                      25000000.0
inTarvo                            21900000.0
Grameen Financial Services         21556050.0